### PR TITLE
[Paimon]support projection for paimon source

### PR DESCRIPTION
**Pull request by TaoZex**
### Purpose of this pull request

support projection for paimon source

### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

### CODE REVIEW

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-paimon-e2e/src/test/resources/paimon_projection_to_assert.conf
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-paimon-e2e/src/test/resources/paimon_projection_to_assert.conf]: Why not use column names but column index？

Comment (TaoZex) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-paimon-e2e/src/test/resources/paimon_projection_to_assert.conf]: I intend to support both. Currently this pr supports column index, and the next pr will support column names.

**Comment by TaoZex** [comment]
PTAL @Hisoka-X @EricJoy2048

**Comment by Hisoka-X** [comment]
cc @dailai

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-paimon-e2e/src/test/resources/paimon_projection_to_assert.conf
Comment (dailai) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-paimon-e2e/src/test/resources/paimon_projection_to_assert.conf]: > I intend to support both. Currently this pr supports column index, and the next pr will support column names.

Usually we use the field name, it is strange to use index, it is recommended to support only the field name.

**Comment by dailai** [comment]
Please support projection by field name.

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-paimon-e2e/src/test/resources/paimon_projection_to_assert.conf
Comment (TaoZex) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-paimon-e2e/src/test/resources/paimon_projection_to_assert.conf]: Thanks. I will modify the code according to your suggestion.

**Comment by TaoZex** [comment]
> Please support projection by field name.

Fixed. Thanks.

**Comment by dailai** [comment]
LGTM

**Review by dailai — APPROVED** [review]

File: docs/en/connector-v2/source/Paimon.md
Comment (Hisoka-X) [inline docs/en/connector-v2/source/Paimon.md]:
```suggestion
### read_columns [string]
```
How about like other connectors, named `read_columns` https://seatunnel.apache.org/docs/connector-v2/source/LocalFile#read_columns-list

**Comment by Hisoka-X** [comment]
Hi @TaoZex , could you solve the conflict?

**Comment by TaoZex** [comment]
> Hi @TaoZex , could you solve the conflict?

Of course.

File: docs/en/connector-v2/source/Paimon.md
Comment (TaoZex) [reply docs/en/connector-v2/source/Paimon.md]: @Hisoka-X @hailin0 I have fixed it as suggested, please review it again.

File: docs/en/connector-v2/source/Paimon.md
Comment (Hisoka-X) [inline docs/en/connector-v2/source/Paimon.md]: please change the document too

Comment (TaoZex) [reply docs/en/connector-v2/source/Paimon.md]: Fixed, thanks.

**Review by Hisoka-X — DISMISSED** [review]

**Comment by Carl-Zhou-CN** [comment]
For https://github.com/apache/seatunnel/pull/6887
There also seems to be an intention to cover the projection, at which point we will have two projection parameters,Do we need to keep both? @Hisoka-X @TaoZex @dailai

**Comment by TaoZex** [comment]
> For #6887 There also seems to be an intention to cover the projection, at which point we will have two projection parameters,Do we need to keep both? @Hisoka-X @TaoZex @dailai

This pr implements the projection for paimon. After the pr was merged, https://github.com/apache/seatunnel/pull/6887 can be further through SQL for the projection.

**Comment by Carl-Zhou-CN** [comment]
> > For #6887 There also seems to be an intention to cover the projection, at which point we will have two projection parameters,Do we need to keep both? @Hisoka-X @TaoZex @dailai
> 
> This pr implements the projection for paimon. After the pr was merged, #6887 can be further through SQL for the projection.

@TaoZex  I think 'read_columns' can be overlaid in 'query sql'. Can we transform this pr based on sql？

**Comment by TaoZex** [comment]
> > > For #6887 There also seems to be an intention to cover the projection, at which point we will have two projection parameters,Do we need to keep both? @Hisoka-X @TaoZex @dailai
> > 
> > 
> > This pr implements the projection for paimon. After the pr was merged, #6887 can be further through SQL for the projection.
> 
> @TaoZex I think 'read_columns' can be overlaid in 'query sql'. Can we transform this pr based on sql？

I had a discussion with @dailai  on wechat. Currently his pr only supports select * sql queries. I will wait for his pr to merge, and modify this pr code to select id, name sql query (Assuming multiple fields, but only two fields are queried by projection)
cc @hailin0 @Hisoka-X

**Comment by hailin0** [comment]
this pr https://github.com/apache/seatunnel/pull/6887 is merged
@TaoZex

**Comment by TaoZex** [comment]
> this pr #6887 is merged @TaoZex

I have changed the code. PTAL. @Hisoka-X @hailin0

**Comment by hailin0** [comment]
cc @dailai

File: seatunnel-connectors-v2/connector-paimon/src/main/java/org/apache/seatunnel/connectors/seatunnel/paimon/source/converter/SqlToPaimonPredicateConverter.java
Comment (dailai) [inline seatunnel-connectors-v2/connector-paimon/src/main/java/org/apache/seatunnel/connectors/seatunnel/paimon/source/converter/SqlToPaimonPredicateConverter.java]: remove this comment

File: seatunnel-connectors-v2/connector-paimon/src/main/java/org/apache/seatunnel/connectors/seatunnel/paimon/source/converter/SqlToPaimonPredicateConverter.java
Comment (dailai) [inline seatunnel-connectors-v2/connector-paimon/src/main/java/org/apache/seatunnel/connectors/seatunnel/paimon/source/converter/SqlToPaimonPredicateConverter.java]: Remove

File: seatunnel-connectors-v2/connector-paimon/src/main/java/org/apache/seatunnel/connectors/seatunnel/paimon/source/PaimonSourceSplitEnumerator.java
Comment (dailai) [inline seatunnel-connectors-v2/connector-paimon/src/main/java/org/apache/seatunnel/connectors/seatunnel/paimon/source/PaimonSourceSplitEnumerator.java:54]: The projection is not used here

**Review by Hisoka-X — DISMISSED** [review]
LGTM. wating @dailai

**Review by dailai — APPROVED** [review]
LGTM @hailin0

**Comment by TaoZex** [comment]
> LGTM @hailin0

Thanks for review. Done.

**Review by hailin0 — APPROVED** [review]
LGTM

**Review by Hisoka-X — APPROVED** [review]

**Pull request merged**
